### ISSUE TITLE
Is it necessary to require the input `log_probs` to CTC loss to be sum-to-one?

### ISSUE THREAD

**Issue by huangruizhe** [pytorch/pytorch#122243]
In pytorch's implementation of CTC loss, the input probability `log_probs` is required to be a valid, sum-to-one probability:

https://github.com/pytorch/pytorch/blob/bc1fef113d425395adfae8a4372f27c68daeb93d/torch/nn/functional.py#L2630-L2653

However, in practice, the probabilities may be modified (e.g., by some sort of scaling, [label priors](https://arxiv.org/abs/2105.14849), etc) so that they don't sum to one anymore. Nevertheless, this still makes sense to compute CTC loss -- which is basically to use dynamic programming to sum over the probabilities (-- which are actually some "scores" that are not necessarily valid probabilities --) corresponding to a given target. 

Given the general application of CTC loss, is it possible to support the non-sum-to-one version of CTC?

Note, in the current version of pytorch's CTC, this is NOT supported yet, as demonstrated in [this google colab notebook](https://colab.research.google.com/drive/1xgsjvjGGXyc4F1vK2bxk8o-BlJ1fJ_Fb?usp=sharing). When feeding pytorch's CTC with non-sum-to-one `log_probs`, the forward pass is correct while the backward pass is not.

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by huangruizhe** [comment]
A follow up on this:

The reason is probably due to that the [pytorch implementation](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LossCTC.cpp#L366) uses the equation (16) in [Alex's CTC paper](https://www.cs.toronto.edu/~graves/icml_2006.pdf), which assumes the input to the ctc_loss function MUST go through a softmax function RIGHT BEFORE ctc. This assumption is not necessarily true, especially when log_prob does not sum to one, right?
At this point, I think it may be more appropriate to use the equation below eq. (15) to compute the gradients wrt to $y_k^t$ instead of $u_k^t$, and let the gradient back propagation over softmax be a seperate process.
If that's the case, the modification of ctc loss seems non-trivial to me. It involves modifying the cpp and cuda codes in several places.

On the other hand, [this pure python library](https://github.com/vadimkantorov/ctc/blob/master/ctc.py#L8) computes CTC loss with its own implementation of dynamic programming. So, when it does "backward" to compute gradients, it always gets the correct gradients wrt the input log_prob, no matter whether the log_prob sums to one or not. 
This library doesn't use eq.(16) to compute the gradients here.

**Comment by mikaylagawarecki** [comment]
Adding `module: autograd` because the issue seems to be that the gradients when `log_probs` do not sum to 1 is incorrect

**Comment by huangruizhe** [comment]
Yes, that's correct.
Adding a link to issue #52241 here, which is very related.
